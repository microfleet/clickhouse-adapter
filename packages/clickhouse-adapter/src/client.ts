import ClickHouse from '@makeomatic/clickhouse'
import { Promise } from 'bluebird'
import { merge } from 'lodash'
import { Writable, Duplex } from 'stream'
import {
  InsertData,
  TableBuilder,
} from './interfaces'

export class ClickhouseClient {
  public static readonly defaultOpts: ClickHouse.Options = {
    host: 'clickhouse',
  }

  public readonly connection: ClickHouse

  public readonly queryAsync: (query: string, options: ClickHouse.QueryOptions) => Promise<any>
  public readonly insertAsync: (dbName: string, insertData: InsertData, options?: ClickHouse.QueryOptions) => Promise<any>

  constructor(options: ClickHouse.Options) {
    this.connection = new ClickHouse(merge({}, ClickhouseClient.defaultOpts, options))
    this.queryAsync = Promise.promisify(this.query, { context: this })
    this.insertAsync = Promise.promisify(this.insert, { context: this })
  }

  public async createTable(builder: TableBuilder): Promise<any> {
    return this.queryAsync(builder.toSql(), { format: 'TabSeparated' })
  }

  public async close(): Promise<void> {
    await this.connection.close()
  }

  public insert<T = any>(dbName: string, insertData: InsertData, cb: ClickHouse.Callback<T>): void
  public insert<T = any>(dbName: string, insertData: InsertData, options: ClickHouse.QueryOptions, cb: ClickHouse.Callback<T>): void
  public insert<T = any>(dbName: string, insertData: InsertData, arg1: ClickHouse.QueryOptions | ClickHouse.Callback<T>, arg2?: ClickHouse.Callback<T>): void {
    const queryOptions: ClickHouse.QueryOptions = typeof arg1 === 'object' ? arg1 : Object.create(null)
    const stream = this.connection.query(
      insertData.query(),
      {
        ...queryOptions,
        format: queryOptions.format || 'JSONEachRow',
        queryOptions: { database: dbName },
      },
      typeof arg1 === 'function' ? arg1 : arg2
    )

    const data = insertData.data()

    data.forEach((item: any) => {
      stream.write(item)
    })

    stream.end()
  }

  public query(query: string, options: ClickHouse.QueryOptions): Duplex
  public query<T = any>(query: string, options: ClickHouse.QueryOptions, cb: ClickHouse.Callback<T>): Writable
  public query<T = any>(query: string, options: ClickHouse.QueryOptions, cb?: ClickHouse.Callback<T>): Writable | Duplex {
    return this.connection.query(
      query,
      { syncParser: true, ...options, format: options.format || 'JSONCompact' },
      cb
    )
  }
}
