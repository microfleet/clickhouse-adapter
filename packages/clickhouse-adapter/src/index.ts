export * from './utils/tableMaker'
export * from './client'
export * from './migrators'
export * from './interfaces'
