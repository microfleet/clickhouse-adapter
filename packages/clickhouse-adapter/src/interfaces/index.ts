export * from './migration'
export * from './insertData'
export * from './structureTable'
export * from './column'
export * from './table-spec'
export * from './clickhouseOptions'
export * from './tableBuilder'
export * from './clickhouseClientInterface'
