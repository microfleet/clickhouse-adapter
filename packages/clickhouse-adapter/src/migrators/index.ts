export * from './migrator'
export * from './system'
