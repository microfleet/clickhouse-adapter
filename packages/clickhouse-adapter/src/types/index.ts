export * from './dateTime'
export * from './enum'
export * from './columnTypes'
