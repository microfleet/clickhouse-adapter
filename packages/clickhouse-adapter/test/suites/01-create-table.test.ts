import { ClickhouseClient, TableMaker } from '../../src'

describe('create table', () => {
  it('create table', async () => {
    const client = new ClickhouseClient({
      host: 'ch1',
    })

    await client.connection.querying('CREATE DATABASE IF NOT EXISTS db_test')

    await client.createTable(
      new TableMaker('db_test', 'test_table', null, {
        columnDefinitions: [
          { name: 'trackDate', type: 'Date' },
          { name: 'trackTimestamp', type: 'DateTime' },
          { name: 'event_type', type: 'String' },
        ],
        tableOptions: ['ENGINE = MergeTree(trackDate, (event_type), 8192)'],
      })
    )

    const descTable = await client.queryAsync('DESCRIBE TABLE db_test.test_table', {
      format: 'JSON',
    })

    expect(descTable.data).toMatchSnapshot()
  })

  it('create replicated table', async () => {
    const client = new ClickhouseClient({
      host: 'ch1',
    })

    await client.connection.querying('CREATE DATABASE IF NOT EXISTS db_test')

    await client.createTable(
      new TableMaker('db_test', 'test_replicated_table', `'{cluster}'`, {
        columnDefinitions: [
          { name: 'trackDate', type: 'Date' },
          { name: 'trackTimestamp', type: 'DateTime' },
          { name: 'event_type', type: 'String' },
        ],
        tableOptions: [
          `ENGINE = ReplicatedMergeTree('/clickhouse/{installation}/{cluster}/tables/{shard}/{database}/{table}', '{replica}')`,
          'PARTITION BY toYYYYMM(trackDate)',
          'ORDER BY (trackTimestamp)',
        ],
      })
    )

    const descTable = await client.queryAsync('DESCRIBE TABLE db_test.test_replicated_table', {
      format: 'JSON',
    })

    expect(descTable.data).toMatchSnapshot()
  })
})
