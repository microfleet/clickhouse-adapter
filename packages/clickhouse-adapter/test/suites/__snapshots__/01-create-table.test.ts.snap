// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create table create replicated table 1`] = `
Array [
  Object {
    "codec_expression": "",
    "comment": "",
    "default_expression": "",
    "default_type": "",
    "name": "trackDate",
    "ttl_expression": "",
    "type": "Date",
  },
  Object {
    "codec_expression": "",
    "comment": "",
    "default_expression": "",
    "default_type": "",
    "name": "trackTimestamp",
    "ttl_expression": "",
    "type": "DateTime",
  },
  Object {
    "codec_expression": "",
    "comment": "",
    "default_expression": "",
    "default_type": "",
    "name": "event_type",
    "ttl_expression": "",
    "type": "String",
  },
]
`;

exports[`create table create table 1`] = `
Array [
  Object {
    "codec_expression": "",
    "comment": "",
    "default_expression": "",
    "default_type": "",
    "name": "trackDate",
    "ttl_expression": "",
    "type": "Date",
  },
  Object {
    "codec_expression": "",
    "comment": "",
    "default_expression": "",
    "default_type": "",
    "name": "trackTimestamp",
    "ttl_expression": "",
    "type": "DateTime",
  },
  Object {
    "codec_expression": "",
    "comment": "",
    "default_expression": "",
    "default_type": "",
    "name": "event_type",
    "ttl_expression": "",
    "type": "String",
  },
]
`;
